### DISCUSSION TITLE
Copy Cef/CefSharp into sub folder, does not work

### DISCUSSION THREAD

**Discussion by trivalik** [cefsharp/CefSharp#3532] [Show and tell]
I converted my project to new .NET Core SDK style projects (still compiling agains .NET Framework 4.8), with use of CefSharp 86, all is fine. CefSharp is in sub folder copied. Also if I build a project without Cefsharp in it but a dependency is using it.
Now I want to update to CefSharp 89, found that building projects which depends on this CefSharp works fine, but projects without direct CefSharp reference (nuget package) which rely on these projects with CefSharp dependency do now copy the cef binaries direct into OutputDir. This is because of the _TransitiveItemsToCopyToOutputDirectoryPreserveNewest --> PreserveNewest lead to copy it to output directory.

I guess it started since CefSharp 87. CefSharp 90.5.70-pre has no real change here, so it should have same issue.

I read about `DisableTransitiveProjectReferences`, but this is not working somehow.

Edit:

One workaround would to define `CefSharpTargetDir` this these projects too, but this will be getting dependency hell.

**Comment by amaitland** [comment]
> Now I want to update to CefSharp 89, found that building projects which depends on this CefSharp works fine, but projects without direct CefSharp reference (nuget package) which rely on these projects with CefSharp dependency do now copy the cef binaries direct into OutputDir.

Issue #3319 has details on the nuget package restructure. In particular https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-818354408

> I read about `DisableTransitiveProjectReferences`, but this is not working somehow.

I'm not familiar with that particular option, please provide a reference link.

> One workaround would to define `CefSharpTargetDir` this these projects too, but this will be getting dependency hell.

The preferred option in my opinion is to use a [Directory.Build.props](
https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2019#directorybuildprops-example) file as I've described in https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-818354408

You can set `<CefSharpTargetDir/>` in a solution wide basis if required. The example I used was for `<CefSharpAnyCpuSupport/>` the same fundamentals apply to `<CefSharpTargetDir/>`.

You can also set `<CefSharpBuildAction />` to `NoAction` for projects you don't want to copy any of the files into output folder, this is particularly useful for `Library` projects where you only require the files in the `Exe/WinExe` projects. Save compile time/disk space.
```xml
<CefSharpBuildAction Condition="'$(OutputType)' == 'Library'">NoAction</CefSharpBuildAction>
```

> which rely on these projects with CefSharp dependency do now copy the cef binaries direct into OutputDir

That is the expected default when using a `Sdk Style` project, the default is that `buildTransitive` assets flow through to consuming projects so reference libraries can consume the projects without requiring a direct reference.

You can set `<CefSharpBuildAction/>` to `None` and control the [Control the dependency assets](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets) using the `<PackageReference/>` element. to effectively revert to the old pre version `87` behavior, though the preferred option in my opinion is a `Directory.Build.props` file.

**Reply by trivalik** [reply]
I have no real reference for `DisableTransitiveProjectReferences`, maybe is it only really working for old style project.

I forgot to tell, that I am targeting only x86 right know.

Adding Directory.Build.props to solution directory to prevent copy (https://curia.me/net-core-transitive-references-and-how-to-block-them/) is not the trick.

I did for every direct reference this:
```
<PackageReference Include="cef.redist.x86" Version="89.0.17">
  <PrivateAssets>all</PrivateAssets>
</PackageReference>
<PackageReference Include="CefSharp.WinForms" Version="89.0.170">
  <PrivateAssets>all</PrivateAssets>
  <ExcludeAssets>runtime</ExcludeAssets>
</PackageReference>
```

`<ExcludeAssets>runtime</ExcludeAssets>` prevent copy to output directory. Then I copied them manually into the sub folder.

Edit:
seems now also working without PackageReference Include="cef.redist.x86

**Reply by amaitland** [reply]
> Adding Directory.Build.props to solution directory to prevent copy (https://curia.me/net-core-transitive-references-and-how-to-block-them/) is not the trick.

What do you mean by this? Did you try it and it didn't work? If so what did you try? Where did you put the file? Please provide detailed responses.

> I did for every direct reference this:

For PackageReference you need only include the top level package, keep CefSharp.Winforms and remove the other CefSharp packages. 

I don't know the exact values you need, you will also need to set `<CefSharpBuildAction/>` to None so the files are included as they were in previous version.

**Reply by trivalik** [reply]
using just this setting for every project which uses CefSharp:
```
  <ItemGroup>
    <PackageReference Include="CefSharp.WinForms" Version="89.0.170" />
  </ItemGroup>
  <PropertyGroup>
    <CefSharpTargetDir>CefSharp</CefSharpTargetDir>
    <CefSharpBuildAction>None</CefSharpBuildAction>
  </PropertyGroup>
```

still copies cef.pak, and co. into target folder. Also CefSharp.WinForms.dll and co. is not inside the `CefSharpTargetDir`.
Anyway, my solution works for now. Thanks for your time. In case you want to improve here, I can test.

**Reply by amaitland** [reply]
> using just this setting for every project which uses CefSharp:

You also need to make buildTransative a private asset.

https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets

> CefSharp.WinForms.dll and co. is not inside the CefSharpTargetDir.

As per https://github.com/cefsharp/CefSharp/issues/3319 that is expected, CefSharp.Winforms.dll, CefSharp.dll and CefSharp.Core.dll now target AnyCpu and are added as direct references to the project by Nuget, they cannot be moved by CefSharpTargetDir. You can add a custom post build action if you absolutely have to move these files.

> In case you want to improve here, I can test.

Did you test with a Directory.Build.props file? What did your file include?

**Comment by trivalik** [comment]
I tried it but without `<ExcludeAssets>runtime</ExcludeAssets>` are files still copied. With my above settings does it work.

**Reply by amaitland** [reply]
Tried what exactly? Please include a detailed example.

**Reply by trivalik** [reply]
With V1 is in "Out" folder additional copied the cef.pak,....  The CefSharp.Winforms.dll are also not in subfolder "CefSharp".
[cefExample_V1.zip](https://github.com/cefsharp/CefSharp/files/6420043/cefExample_V1.zip)

Adding:

```
<PackageReference Include="CefSharp.WinForms" Version="89.0.170">
  <PrivateAssets>all</PrivateAssets>
  <ExcludeAssets>runtime</ExcludeAssets>
</PackageReference>
```

removed this in V2 example, the `<CefSharpBuildAction>None</CefSharpBuildAction>` makes no difference.

[cefExample_v2.zip](https://github.com/cefsharp/CefSharp/files/6420064/cefExample_v2.zip)

**Reply by amaitland** [reply]
> The CefSharp.Winforms.dll are also not in subfolder "CefSharp".

As I've already said, this is expected. 

As a general rule I don't open zip files from unknown sources. If you upload your examples to GitHub I'll have a look.

**Reply by trivalik** [reply]
The zip file source is github :-), or should I create a repo?

The copy behavior of CefSharp.Winforms.dll was in Version 86 different, then there was it copied to this subfolder.

**Reply by amaitland** [reply]
> or should I create a repo?

Yes please.

> The copy behavior of CefSharp.Winforms.dll was in Version 86 different, then there was it copied to this subfolder.

Correct. That behaviour has now changed. There is now only a single `AnyCPU` version of `CefSharp.WinForms.dll` so there is no need to include it based on architecture, so now `Nuget` includes the dll like any other.

https://github.com/cefsharp/CefSharp/blob/cefsharp/89/NuGet/CefSharp.WinForms.nuspec#L30

**Reply by trivalik** [reply]
you can find if necessary at https://github.com/trivalik/cefExampleDiscussion

**Reply by amaitland** [reply]
> you can find if necessary at https://github.com/trivalik/cefExampleDiscussion

Thanks. I've had a look and have updated your examples demonstrating the two options I've been suggesting, both work as expected using VS2019 (16.8.6).

Option 1 - Make `buildTransitive` a PrivateAsset and set `<CefSharpBuildAction>None</CefSharpBuildAction>`.

Changes made in commit https://github.com/amaitland/cefExampleDiscussion/commit/ff57edf22ac86c79c96513d72270e36faa7deadb

Project file at https://github.com/amaitland/cefExampleDiscussion/blob/ff57edf22ac86c79c96513d72270e36faa7deadb/cefExample_V1/CefUsingLib/CefUsingLib.csproj

Option 2 - Use a Directory.Build.props to configure CefSharpTargetDir

Changes made in commit https://github.com/amaitland/cefExampleDiscussion/commit/989314fe464ef6c34b0842dd3b3d712b45ca4288

Both options give the same output.

**Reply by trivalik** [reply]
I want now to get rid of the AnyCPU dlls copy by my self into `CefSharpTargetDir`.

I tried yet only option 1 with CefSharp 90. I put
```
    <PackageReference Include="CefSharp.WinForms" Version="90.6.50">
      <PrivateAssets>buildTransitive</PrivateAssets>
    </PackageReference>
```
in every project which reference CefSharp.WinForms. Additional I tried to add `CefSharpBuildAction`, but if present or not did not any difference. What I found, as for all nuget packages, that clean does not remove the CefSharp.dll, CefSharp.WinForms.dll and CefSharp.Core.dll. Also the CefSharp.dll is present inside the `CefSharpTargetDir` and also outside in the `OutputPath`. Also the line `<CefSharpBuildAction Condition="'$(OutputType)' == 'Library'">NoAction</CefSharpBuildAction>` brings no difference.

**Reply by amaitland** [reply]
> I want now to get rid of the AnyCPU dlls copy by my self into CefSharpTargetDir.

If by get rid of you mean move the dlls then you'll have to add a post build task. 

> What I found, as for all nuget packages, that clean does not remove the CefSharp.dll, CefSharp.WinForms.dll and CefSharp.Core.dll. 

These files are included as libraries, they are now AnyCPU dlls. Changing buildTransitive to private will have no impact on these files. They are runtime/runtime assets, please make sure you've read https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets

> Also the CefSharp.dll is present inside the CefSharpTargetDir

The CefSharp.BrowserSubProcess.exe requires CefSharp.dll so an extra copy is added to allow it to load. This is required.

> Also the line <CefSharpBuildAction Condition="'$(OutputType)' == 'Library'">NoAction</CefSharpBuildAction> brings no difference.

You must use this in conjunction with a Directory.Build.props

**Reply by trivalik** [reply]
With AnyCPU I mean that I take over your separation. That AnyCPU DLLs are not part of x86 folder, except CefSharp.dll.

With your `buildTransitive` seems work, but has same issue as mine in case an consuming project which reference `CefSharp.WinForms.dll` wants to change `CefSharpTargetDir` in respect to an dependency with `CefSharp.WinForms.dll`. This is necessary for unit testing.

**Reply by amaitland** [reply]
> With your buildTransitive seems work, but has same issue as mine in case an consuming project which reference CefSharp.WinForms.dll wants to change CefSharpTargetDir in respect to an dependency with CefSharp.WinForms.dll. This is necessary for unit testing.

I don't understand what you mean by this.

**Reply by trivalik** [reply]
You can see now in https://github.com/trivalik/cefExampleDiscussion/blob/master/cefExample_V3/UnitTestWithCefLibrary/UnitTestWithCefLibrary.csproj this case that UnitTestWithCefLibrary will copy Cef double. I solved this right now with deletion.

**Reply by amaitland** [reply]
> will copy Cef double

What files exactly are you referring? 

Please confirm you've read https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets

**Reply by trivalik** [reply]
I read this. It copied the whole Cef folder twice, because of different path of `CefSharpTargetDir`. To fix this without remove, I have probably to add CefSharp.Common dependency to disable this copy. Anyway, to keep it simple I let the remove directory in.

**Reply by amaitland** [reply]
Is there a reason you don't use a Directory.Build.props file to set CefSharpTargetDir for all your projects?

**Reply by trivalik** [reply]
I did, but the test project have to emulate the same directory structure, else the test would fail.
Maybe I missed in example to move it to root folder, but this would make no difference.

**Reply by Xan-Kun** [reply]
I tried the way given in https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-818354408 and others and none of them work. The files are copied to the bin folder at publish every time.

Is there a working example using Directory.Build.props AND CefSharp.WinForms.**NETCore** somewhere that I might have missed?

**Comment by Xan-Kun** [comment]
I have a `WinForm` project using `CefSharp.WinForms.NETCore` (124.3.80).
The target is `win-x64` with `net8.0-windows7.0`.

I can't get it to copy to a subfolder. And I am a little confused about what the correct way in this case would be.
In the end, I need to get `libGLESv2.dll`  _out of the way_ because it causes problems with other packages.

Any help/hint much appreciated.

**Reply by trivalik** [reply]
I guess you have then to disable the copy by CefSharp packages and manage this manuell file for file.

**Reply by Xan-Kun** [reply]
I dont know how to disable the copy, but i understand the idea.

**Reply by amaitland** [reply]
> I have a `WinForm` project using `CefSharp.WinForms.NETCore` (124.3.80).

`CefSharpTargetDir` is not supported by the `NETCore` packages. The packages use [Architecture-specific folders](https://learn.microsoft.com/en-us/nuget/create-packages/supporting-multiple-target-frameworks#architecture-specific-folders) to comply with the nuspec standard. 

When the `.Net Framework` packages were created, this feature wasn't available, so we copied the files manually, `CefSharpTargetDir` allowed for customisation of this copy. The `NETCore` packages rely on `MSBuild` copy all files (except locales). This allows us to use built in features like [Controlling dependency assets](https://learn.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets)

> I dont know how to disable the copy, but i understand the idea.

See https://github.com/NuGet/Home/issues/5986

**Reply by Xan-Kun** [reply]
Thanks for the comment, I know those already, but they dont really apply. As I understand it, it allows for control on a pe-file basis. (Please correct me if i'm wrong), which is kinda the opprosite of what CefSharpTargetDir was doing.
This approach seems very tideous and error-prone.
How could we reliable tell the difference between the managed dlls, that we need in the binary folder and the "extra" files only for the browser subprocess?

**Comment by Xan-Kun** [comment]
I think it would be fair to answer at least one of the many questions on the web concerning this with a truthful: NO.
Not in CefSharp.WinForms.NETCore.
